packages/models/src/errors.ts

```diff
@@ -144,16 +163,24 @@ export function makeApiProblemBuilder<T extends string>(
     if (operationalLogMessage) {
       logger.warn(operationalLogMessage);
     }
+
     return match<unknown, Problem>(error)
       .with(P.instanceOf(ApiError<T | CommonErrorCodes>), (error) => {
         const mappedCode = httpMapper(error);
-        const code =
-          mappedCode === HTTP_STATUS_INTERNAL_SERVER_ERROR
-            ? defaultCommonErrorMapper(error.code as CommonErrorCodes)
-            : mappedCode;
+        const isInternalServerError =
+          mappedCode === HTTP_STATUS_INTERNAL_SERVER_ERROR;
+
+        const code = isInternalServerError
+          ? defaultCommonErrorMapper(error.code as CommonErrorCodes)
+          : mappedCode;
 
         const problem = makeProblem(code, error);
         logger.warn(makeProblemLogString(problem, error));
+
+        if (hideInternalErrorDetails && isInternalServerError) {
+          return genericProblem;
```

Comment:
Maybe we could add a log here that makes clear that the error we just logged is not returned, and a generic problem is returned instead?
Otherwise, the log above does not correspond to the returned error, and it's hard to understand why.
Something like the following (or maybe you have a better proposal...)

```suggestion
          logger.info("forceGenericProblemOn500 flag is true: returning generic problem instead of the error logged above")
          return genericProblem;
```